### PR TITLE
DO NOT MERGE! AES256CounterRNG built-in replacement (#12636)

### PR DESCRIPTION
**Pull request by gosubpl**
This PR is intended to discuss the possible replacement of AES256CounterRNG by alternative built-in function doing the same.

### CODE REVIEW

**Comment by gosubpl** [comment]
Refs #12636

**Comment by gosubpl** [comment]
@patriknw @drewhk would be grateful if you took a look at it.
This is an incomplete code - for discussion only - **please do not merge** (yet)

AESCounterBuiltinRNG is a fresh implementation in Scala of the AESCounterRNG algorithm from uncommons.math. The whole library is Apache License 2.0 and the AES CTR RNG is a folk-lore algorithm widely that is widely used.

It is accompanied by AES256CounterBuiltinRNG which is a copy of AES256CounterRNG wrapping the new algorithm (AESCounterBuiltinRNG) instead of the old one (AES256CounterRNG). It also contains direct reference to Java's SecureRandom seed generation instead of going through uncommons.math wrapper.

Two accompanying tests show that:

1. The new code generates the same random numbers as the old one
2. The performance of the new code is similar to old one (I would say around 10-15% worse - see copied example runs of jmh benchmar):
```
akka-bench-jmh > jmh:run -wi 15 -i 15 -f 1 .*RandomReplacementBenchmark.*
[...]
[info] Benchmark                          Mode  Cnt    Score    Error  Units

[info] RandomReplacementBenchmark.runNew    ss   15  254.022 ± 31.966  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  261.415 ± 36.318  ms/op

[info] RandomReplacementBenchmark.runNew    ss   15  372.153 ± 23.552  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  306.477 ± 32.487  ms/op

[info] RandomReplacementBenchmark.runNew    ss   15  249.846 ± 28.165  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  210.568 ± 14.917  ms/op

[info] RandomReplacementBenchmark.runNew    ss   15  373.674 ±  85.931  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  384.629 ± 101.545  ms/op

[info] RandomReplacementBenchmark.runNew    ss   15  249.659 ± 20.295  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  262.815 ± 49.112  ms/op
```

Possible strategies of going forward are:
1. Add AESCounterBuiltinRNG together with AES256CounterBuiltinRNG and AES128CounterBuiltinRNG to codebase, add them also to the list of available codes and deprecate the AES256CounterRNG (together with AES128CounterRNG). To remove uncommons.math at later time remove the the already deprecated Inet implementations and the non-builtin implementations.
2. Replace AESCounterRNG with AESCounterBuiltinRNG in AES256CounterRNG (and its 128 bit counterpart) - this will effectively replace the AES256CounterRNG implementation with the AES256CounterBuiltinRNG implementation. After that to remove uncommons.math just remove the already deprecated Inet RNG. Will be binary incompatible but the protocol will match (as the generated numbers are the same)

I would suggest going forward with 2.

File: akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  private val COUNTER_SIZE_BYTES = 16
+  private val BITWISE_BYTE_TO_INT = 0x000000FF
+
+  // mutable state below, concurrent accesses need lock
+  private val lock = new ReentrantLock
+  private val counter: Array[Byte] = new Array[Byte](COUNTER_SIZE_BYTES)
+  private var index: Int = 0
+  private var currentBlock: Array[Byte] = null
+
+  private val cipher = Cipher.getInstance("AES/ECB/NoPadding")
```
Comment (drewhk) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala]: wait, doesn't Cipher has an AES/CTR/NoPadding anyway? Why hand-build the counter? CTR just XORs the plaintext with the current counter output. I.e. you should just continuously encrypt the all-zero block and take its output as the random value. Also, I would prefer if this is automatically reseeded from time to time from the SecureRandom, i.e. after a certain number of outputs.

Comment (gosubpl) [inline akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala]: I have added `AESCounterBuiltinCTRRNG` that is using `AES/CTR/NoPadding`. To allow verification I have added custom iv with value of 1 but, obviously, this can be removed before merge.
Works great, only issue being that it is even slower, as the zero block needs now to be XORed with the encrypted counter. So the performance looks like that:
```
[info] Benchmark                          Mode  Cnt    Score    Error  Units
[info] RandomReplacementBenchmark.runCtr    ss   15  299.485 ± 23.466  ms/op
[info] RandomReplacementBenchmark.runNew    ss   15  238.772 ± 12.127  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  215.453 ± 24.637  ms/op

[info] RandomReplacementBenchmark.runCtr    ss   15  283.690 ± 17.258  ms/op
[info] RandomReplacementBenchmark.runNew    ss   15  251.412 ± 25.066  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  231.133 ± 29.139  ms/op

[info] RandomReplacementBenchmark.runCtr    ss   15  314.895 ± 45.711  ms/op
[info] RandomReplacementBenchmark.runNew    ss   15  250.495 ± 15.762  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  215.201 ± 23.651  ms/op
```
which means that the CTR algo is **1.5x** slower than the original one. If that is fine, great, I will happily replace that as hand-built counter is also hurting my eyes.
Of course, aye to reseeding. This is missing from the original implementation, and, for realistic messaging volumes, badly needed. I will add this as soon as we settle the algorithm.

**Comment by gosubpl** [comment]
Copying the comment here, as it is not shown by default for the "outdated" block.

I have added `AESCounterBuiltinCTRRNG` that is using `AES/CTR/NoPadding`. To allow verification I have added custom iv with value of 1 but, obviously, this can be removed before merge.
Works great, only issue being that it is even slower, as the zero block needs now to be XORed with the encrypted counter. So the performance looks like that:
```
[info] Benchmark                          Mode  Cnt    Score    Error  Units
[info] RandomReplacementBenchmark.runCtr    ss   15  299.485 ± 23.466  ms/op
[info] RandomReplacementBenchmark.runNew    ss   15  238.772 ± 12.127  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  215.453 ± 24.637  ms/op

[info] RandomReplacementBenchmark.runCtr    ss   15  283.690 ± 17.258  ms/op
[info] RandomReplacementBenchmark.runNew    ss   15  251.412 ± 25.066  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  231.133 ± 29.139  ms/op

[info] RandomReplacementBenchmark.runCtr    ss   15  314.895 ± 45.711  ms/op
[info] RandomReplacementBenchmark.runNew    ss   15  250.495 ± 15.762  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  215.201 ± 23.651  ms/op
```
which means that the CTR algo is **1.5x** slower than the original one. If that is fine, great, I will happily replace that as hand-built counter is also hurting my eyes.
Of course, aye to reseeding. This is missing from the original implementation, and, for realistic messaging volumes, badly needed. I will add this as soon as we settle the algorithm.

**Comment by johanandren** [comment]
Relevant for this PR, is the comment Will made on the SSL config doc ticket: https://github.com/akka/akka/pull/21730#discussion_r85265351

**Comment by gosubpl** [comment]
@johanandren yes, it is. Please see performance comparison below, `SecureRandom` is like **12x** slower.
```
[info] Benchmark                             Mode  Cnt     Score     Error  Units
[info] RandomReplacementBenchmark.runCtr       ss   15   297.946 ±  39.956  ms/op
[info] RandomReplacementBenchmark.runNew       ss   15   264.695 ±  35.462  ms/op
[info] RandomReplacementBenchmark.runOld       ss   15   275.792 ±  63.000  ms/op
[info] RandomReplacementBenchmark.runSecRnd    ss   15  3143.013 ± 378.316  ms/op

[info] Benchmark                             Mode  Cnt     Score     Error  Units
[info] RandomReplacementBenchmark.runCtr       ss   15   335.211 ±  77.456  ms/op
[info] RandomReplacementBenchmark.runNew       ss   15   261.172 ±  48.231  ms/op
[info] RandomReplacementBenchmark.runOld       ss   15   246.999 ±  36.847  ms/op
[info] RandomReplacementBenchmark.runSecRnd    ss   15  3172.568 ± 449.505  ms/op
```

I believe that when I first suggested in akka/dev gitter channel to just go with `SecureRandom`, @patriknw opposed, saying (if I correctly recall) that we need something faster. If that is the case, from this comparison I would lean towards @drewhk suggestion to use straight `AES-CTR` mode, as some of its relative slowness would probably come from additional xoring and (hopefully) not ending the "add one to CTR" when there is nothing more to carry which could be susceptible to timing attacks. Other options would be [Fortuna](https://en.wikipedia.org/wiki/Fortuna_(PRNG)) - which practically is AES-CTR with rekeying, but it would have to be probably well tested according to [NIST SP 800-22](http://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-22r1a.pdf) to avoid errors in the implementation (I think that uncommons.math's PRNGs were statistically tested).
Basically if you think `SecureRandom` (straight /dev/urandom, according to @wsargent post) is the way to go, all PRNGs from `uncommons.math` should be deprecated and this PR close. Then, at the next opportunity the whole library can be removed. Otherwise, please let me know which algo you would prefer to implement.

**Comment by wsargent** [comment]
Okay, that's interesting.  It makes sense that even a straight call to `/dev/urandom` is slower than a CSPRNG.  It looks like the benchmark is directly against the algorithm though -- is there a comparative test with JSSE with different RNG?

**Comment by gosubpl** [comment]
@wsargent will try to cobble something up, as I am also curious what about other RNGs :)

**Comment by wsargent** [comment]
Could this be related to https://bugs.openjdk.java.net/browse/JDK-8098581 ?

**Comment by gosubpl** [comment]
Hmmm... maybe, but I wouldn't exactly call my request _small sized_:
```
inputSecRnd = Array.fill[Byte](10000000)(1)
```
My way of thinking was that PRNGs are faster than "real" RNGs due to high level of re-use of the original randomness. Need constant re-seeding, though, as the entropy decays.

**Comment by gosubpl** [comment]
Closing as I have this mistakenly rebased on recent changes in master. Will reopen again.

**Comment by gosubpl** [comment]
Discussion moved to https://github.com/akka/akka/issues/21740 . Please note that the tests in this PR were done on OpenJDK.

**Comment by drewhk** [comment]
@wsargent of course urandom is slower, there is lot's of overhead accessing files, even if they are not backed by disc compared to calling into AES directly (which has some hardware acceleration on recent JDKs although not great).
